packages/contracts/test/utils/test_with_reference.ts

```diff
@@ -0,0 +1,98 @@
+ * or if the reference function throws/rejects, an "expected error". Next, the
+ * test function will be called to obtain an "actual result", or if the test
+ * function throws/rejects, an "actual error". The test passes if at least one
+ * of the following conditions is met:
+ *
+ * 1) Neither the reference function or the test function throw and the
+ * "expected result" equals the "actual result".
+ *
+ * 2) Both the reference function and the test function throw and the "actual
+ * error" message *contains* the "expected error" message.
+ *
+ * @param referenceFuncAsync a reference function implemented in pure
+ * JavaScript/TypeScript which accepts N arguments and returns the "expected
+ * result" or throws/rejects with the "expected error".
+ * @param testFuncAsync a test function which, e.g., makes a call or sends a
+ * transaction to a contract. It accepts the same N arguments returns the
+ * "actual result" or throws/rejects with the "actual error".
+ * @param values an array of N values, where each value corresponds in-order to
+ * an argument to both the test function and the reference function.
+ * @return A Promise that resolves if the test passes and rejects if the test
+ * fails, according to the rules described above.
+ */
+export async function testWithReferenceFuncAsync(
+    referenceFunc: (...args: any[]) => Promise<any>,
+    testFuncAsync: (...args: any[]) => Promise<any>,
+    values: any[],
+): Promise<void> {
+    const paramNames = _getParameterNames(referenceFunc);
+    const testCaseString = JSON.stringify(_.zipObject(paramNames, values));
+    let expectedResult: any;
+    let expectedErr: string | undefined;
+    try {
+        expectedResult = await referenceFunc(...values);
+    } catch (e) {
+        expectedErr = e.message;
+    }
+    try {
+        const actualResult = await testFuncAsync(...values);
+        if (!_.isUndefined(expectedErr)) {
+            throw new Error(
+                `Expected error containing ${expectedErr} but got no error\n\tTest case: ${testCaseString}`,
+            );
+        }
+        expect(JSON.stringify(actualResult)).to.equal(JSON.stringify(expectedResult));
+    } catch (e) {
+        if (_.isUndefined(expectedErr)) {
+            throw new Error(`${e.message}\n\tTest case: ${testCaseString}`);
+        } else {
+            expect(e.message).to.contain(expectedErr, `${e.message}\n\tTest case: ${testCaseString}`);
+        }
+    }
+}
+
+function _getParameterNames(func: (...args: any[]) => any): string[] {
+    return _.toString(func)
+        .replace(/[/][/].*$/gm, '') // strip single-line comments
+        .replace(/\s+/g, '') // strip white space
+        .replace(/[/][*][^/*]*[*][/]/g, '') // strip multi-line comments
+        .split('){', 1)[0]
+        .replace(/^[^(]*[(]/, '') // extract the parameters
+        .replace(/=[^,]+/g, '') // strip any ES6 defaults
+        .split(',')
+        .filter(Boolean); // split & filter [""]
```

Comment:
Did you write this or did you get it from somewhere online? If online, please add a `// Source: ` comment above it.

Reply:
This looks like quite a hack. I would be okay with outputting an tuple/array of values without names and removing this hack.

Reply:
@Recmo I actually found it really hard to understand a test failure when all I could see was an array of 3-4 numbers without any additional context. This is a hack, but it has real utility, is only for testing, and nothing bad is really going to happen if it breaks. It's pretty well isolated too. Okay if we keep it?